_scripts/evaluate_overall_performance.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+os.chdir('/app')
+
+from langsmith import Client
+from langsmith.evaluation import evaluate
+from langsmith.evaluation import LangChainStringEvaluator
+from langsmith.schemas import Run, Example
+from langchain_openai import ChatOpenAI
+
+from croptalk.model_openai_functions import initialize_model
+from dotenv import load_dotenv
+
+load_dotenv('secrets/.env.secret')
+load_dotenv('secrets/.env.shared')
+
+# Define the dataset name and output directory
+DATASET_NAME = "sp-cp-bp-cih-sql-89q"
+OUTPUT_DIR = "data/experiment_results"
+
+# Initialize a model that will be used for evaluation of the qa
+eval_llm = ChatOpenAI(model="gpt-3.5-turbo", streaming=False, temperature=0.0)
+
+client = Client()
+
+# == CORRECTNESS EVALUATOR ==
+def retrieve_question_answer(run: Run, example: Example) -> dict:
+    parsed = {
+        "prediction": run.outputs.get("output"),
+        "reference": example.outputs.get("Answer"),
+        "input": example.inputs.get("Question"),
+    }
+    return parsed
+
+qa_evaluator = LangChainStringEvaluator(
+    "qa",
+    prepare_data=retrieve_question_answer,
+    config={"llm": eval_llm},
+)
+
+# == DOC RETRIEVAL EVALUATOR ==
+def doc_evaluator(run: Run, example: Example) -> dict:
+    """Calculate the exact match score of the run."""
+    expected = example.outputs.get("Doc Name")
+    predicted = run.outputs.get("output")
+    return {"score": expected.lower() in predicted.lower() if expected else None,
+            "key": "contains_doc"}
+
+# == TOOL EVALUATOR ==
+def tool_evaluator(run: Run, example: Example) -> dict:
+    
+    # Accessing the `intermediate_steps`
+    #TODO: Parse the intermediate steps into a list of objects
+    # For now, we just convert the intermediate steps to a string
+    intermediate_steps = run.outputs.get('intermediate_steps')
+    intermediate_steps_str = str(intermediate_steps)
+
+    expected_tool_name = example.outputs.get("Tool")
+    expected_tool = f"tool='{expected_tool_name}'"
```

Comment:
we're searching for "tool=<tool_name>" in intermediate_steps...
but I see "- tool: FindDocs" in langsmith... which makes me curious: how/why does it work??

![image](https://github.com/cropguard-ai/chat-langchain/assets/147736401/91123cb5-717f-40b2-87d3-cc140a87689e)

Reply:
LangSmith returns the tool executed by the agent

Reply:
let's discuss quickly 'cause it's still not clear to me... the test at L60 is if the string expected_tool.lower() in intermediate_steps_str.lower()...